configfile: "workflow/config.yaml"
rule all:
    input:
        #expand("fastqs/{accession}.fastq.gz", accession=config["SRR"]),
        #expand("trimmed/{accession}_trimmed.fq.gz", accession=config["SRR"]),
        #expand("mapped/{accession}.sam", accession=config["SRR"]),
        #expand("mapped/{accession}.bam", accession=config["SRR"]),
        #"ressources/genomes/reference.gtf",
        #expand("index/aureusIndex.{suffix}.bt2", suffix=config["AUREOS"]),
        #expand("count/{accession}.txt", accession=config["SRR"]),
        "results/plot2.pdf",
        "results/plot1.pdf"
        
rule telechargement_sequences:
    output:
        "fastqs/{sracode}.fastq"
    container: 
        "~/ReproHackaton_2023/Docker/sratoolkit/sratoolkit_3.0.7.sif"
    shell:
        "cd fastqs && fasterq-dump {wildcards.sracode}"

rule compresion_sequences:
    input: 
        "fastqs/{sracode}.fastq"
    output:
        "fastqs/{sracode}.fastq.gz"
    container: 
        "~/ReproHackaton_2023/Docker/sratoolkit/sratoolkit_3.0.7.sif"
    shell:
        "gzip {input}"

rule bowtie_index:
    input: 
        "ressources/genomes/reference.fasta"
    output: 
        "index/aureusIndex.{suffix}.bt2"
    container: 
        "docker://nanozoo/bowtie2"
    shell:
        "bowtie2-build {input} index/aureusIndex"

rule trimming:
    input: 
        "fastqs/{sracode}.fastq.gz"
    output: 
        "trimmed/{sracode}_trimmed.fq.gz"
    container: 
        "~/ReproHackaton_2023/Docker/trim_galore/trim_galore_0.6.10.sif"
    shell:
        "trim_galore -q 20 --phred33 --length 25 --output_dir trimmed {input}"

rule mapping:
    input: 
        "trimmed/{sracode}_trimmed.fq.gz"
    output: 
        "mapped/{sracode}.sam"
    container: 
        "docker://nanozoo/bowtie2"
    shell:
        "bowtie2 -x index/aureusIndex -U {input} -S {output}"


rule samTobam:
    input: 
        "mapped/{sracode}.sam"
    output: 
        bam = "mapped/{sracode}.bam",
        bamSort = "mapped/sort_{sracode}.bam",
        bamIndex = "mapped/sort_{sracode}.bam.bai"
    container:
        "docker://staphb/samtools"
    shell:
        ("samtools view -bS -o {output.bam} {input} \
        && samtools sort -O bam -o {output.bamSort} {output.bam} \
        && samtools index {output.bamSort}")
rule gffTogtf:
    input: 
        "ressources/genomes/reference.gff"
    output: 
        "ressources/genomes/reference.gtf"
    container: 
        "docker://bschiffthaler/gffread"
    shell:
        "gffread {input} -T -o {output}"

rule counting:
    input: 
        ref = "ressources/genomes/reference.gtf",
        bam = "mapped/sort_{sracode}.bam"
    output: 
        "count/{sracode}.txt"
    container: 
        "docker://genomicpariscentre/subread"
    shell:
        "featureCounts -p -O -a {input.ref} -o {output} {input.bam}"
    
rule analyse_stat:
    input:
        expand("count/{accession}.txt", accession=config["SRR"]),
        "ressources/genomes/translation.txt"
    output:
        "results/plot2.pdf",
        "results/plot1.pdf"
        #"results/heatmap.pdf"
    container:
        "~/ReproHackaton_2023/Docker/deseq2/deseq2_r_4.3.0.sif"
    script:
        "script.R"